content/posts/2020/2020-02-19-dask-clusters-overview.md

```diff
@@ -0,0 +1,337 @@
+
+In this paradigm once a cluster is set up folks may start their Python session, connect their client to this existing cluster, do some work and disconnect again. They might later come back to that cluster and run further work. The cluster will sit idle in the meantime.
+
+It is also common in this paradigm for multiple users to share this single cluster, however this is not recommended as the Dask scheduler does not manage users or clients separately and work will be executed on a first come first served bases. Therefore we recommend that users use a cluster one at a time.
+
+### Ephemeral clusters
+
+An ephemeral cluster is one which only exists for the duration of the work. In this case a user may SSH onto the machines, run the commands to set up the cluster, connect a client and perform work, then disconnect and exit the Dask processes. A basic way of doing this would be to create a bash script which calls `ssh` and sets up the cluster. You would run this script in the background while performing your work and then kill it once you are done. We will cover other implementations of this in the coming sections.
+
+Ephemeral clusters allow you to leverage a bunch of machines but free them up again when you are done. This is especially useful when you are using a system like a cloud service or a batch scheduler where you have limited credits, or are charged for provisioned resources.
+
+## Adaptivity
+
+Ephemeral clusters are also generally easier to scale as you will likely have an automated mechanism for starting workers. The Dask scheduler maintains an estimate of how long it expects the outstanding work to take to complete. If the scheduler has a mechanism for starting and stopping workers then it will scale up the workers in an attempt to complete all outstanding work within 5 seconds. This is referred to as adaptive mode.
+
+The mechanisms for starting and stopping workers are added via plugins. Many of the implementations we are about to discuss include this logic.
+
+## Connectivity
+
+Dask uses TCP to communicate between client, scheduler and workers by default. This means that all of these components must be on a TCP/IP network with open routes between the machines. Many connectivity problems step from firewalls or private networks blocking connections between certain components. An example of this would be running Dask on a cloud platform like AWS, but running the Python session and client on your laptop while using the free wifi in a coffee shop. You must ensure you are able to route traffic between components, either by exposing the Dask cluster to the internet or by connecting your laptop to the private network via a VPN or tunnel.
```

Comment:
Would it be useful here to specify the *direction* of connectivity - from workers to scheduler only, rather than scheduler to workers (I think, at least!). It might be too much detail for a blog post though.

Reply:
Yeah that's a really good point! Although am I right in thinking that connectivity happens in both directions?

- Worker connects to the scheduler at scheduler address and provides info about worker
- Scheduler then opens an RPC connection back to the worker on the address provided

Reply:
Yeah, writing this comment made me realize I don't actually know! :D

Reply:
Connectivity happens in both directions